addons/controllers/clusterbootstrap_controller.go

```diff
@@ -300,79 +342,371 @@ func (r *ClusterBootstrapReconciler) createOrPatchclusterBootstrapFromTemplate(
 		}
-		provider, err := r.dynamicClient.Resource(*gvr).Namespace(namespace).Get(r.context, pkg.ValuesFrom.ProviderRef.Name, metav1.GetOptions{}, "status")
+		pkgi.Spec.Values = []kapppkgiv1alpha1.PackageInstallValues{
+			{SecretRef: &kapppkgiv1alpha1.PackageInstallValuesSecretRef{
+				Name: secretName},
+			},
+		}
+		return nil
+	}
 
-		if err != nil {
-			log.Error(err, "unable to fetch provider", "provider", pkg.ValuesFrom.ProviderRef, "gvr", gvr)
-			return nil, err
+	_, err = controllerutil.CreateOrPatch(r.context, r.Client, pkgi, pkgiMutateFn)
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("unable to create or patch %s PackageInstall CR for cluster: %s",
+			constants.KappControllerAddonName, cluster.Name))
+		return err
+	}
+
+	return nil
+}
+
+// createOrPatchPackageOnRemote creates the Package CR on remote cluster. In order to install a package on remote cluster
+// the Package CR needs to be present.
+// createOrPatchPackageOnRemote returns a tuple: (<remote-package>, <error>)
+func (r *ClusterBootstrapReconciler) createOrPatchPackageOnRemote(cluster *clusterapiv1beta1.Cluster,
+	cbPkg *runtanzuv1alpha3.ClusterBootstrapPackage, clusterClient client.Client) (*kapppkgv1alpha1.Package, error) {
+
+	var err error
+	// Create or patch Package CR on remote cluster
+	localPackage := &kapppkgv1alpha1.Package{}
+	key := client.ObjectKey{Namespace: cluster.Namespace, Name: cbPkg.RefName}
+	if err = r.Client.Get(r.context, key, localPackage); err != nil {
+		// If there is an error to get the Carvel Package CR from local cluster, nothing needs to be created/cloned on remote.
+		// Let the reconciler try again.
+		r.Log.Error(err, fmt.Sprintf("unable to create or patch Package %s on cluster %s/%s. Error occurs when getting Package %s from the management cluster",
+			key.String(), cluster.Namespace, cluster.Name, key.String()))
+		return nil, err
+	}
+	remotePackage := &kapppkgv1alpha1.Package{}
+	remotePackage.SetName(localPackage.Name)
+	// The Package CR on remote cluster needs to be under configured system namespace
+	remotePackage.SetNamespace(r.Config.SystemNamespace)
+	_, err = controllerutil.CreateOrPatch(r.context, clusterClient, remotePackage, func() error {
+		remotePackage.Spec = *localPackage.Spec.DeepCopy()
+		// TODO: Follow up to see if we need to preserve all the other fields, like annotations
+		// https://github.com/vmware-tanzu/tanzu-framework/issues/1678
+		return nil
+	})
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("unable to create or patch Package resource %s/%s on cluster: %s/%s",
+			remotePackage.Namespace, remotePackage.Name, cluster.Namespace, cluster.Name))
+		return nil, err
+	}
+	return remotePackage, nil
+}
+
+// createOrPatchPackageInstallOnRemote creates or patches PackageInstall CR on remote cluster. The kapp-controller
+// running on remote cluster will reconcile it and deploy resources.
+func (r *ClusterBootstrapReconciler) createOrPatchPackageInstallOnRemote(cluster *clusterapiv1beta1.Cluster,
```

Comment:
Don't you need to check for ```constants.ManagementClusterRoleLabel``` to not end up installing the package in the management cluster here?

Reply:
we only skip kapp-controller installation on mgmt cluster. There is an check at L365. Other packages are following the same pattern to be installed on mgmt cluster.